File: backend/app/routers/metadata.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.auth.dependencies import get_current_active_user
from app.core.responses import ApiResponse
from app.db.session import get_db
from app.schemas import MetadataCreate, MetadataRead, MetadataUpdate
from app.services.metadata_service import MetadataService


router = APIRouter(
    prefix="/metadata",
    tags=["metadata"],
    dependencies=[Depends(get_current_active_user)],
)


@router.post("", response_model=ApiResponse[MetadataRead])
def create_metadata(
    payload: MetadataCreate,
    db: Session = Depends(get_db),
) -> ApiResponse[MetadataRead]:
    metadata = MetadataService(db).create(payload)
    return ApiResponse(message="Metadata created.", data=metadata)


@router.get("", response_model=ApiResponse[list[MetadataRead]])
def list_metadata(
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=100, ge=1, le=500),
    db: Session = Depends(get_db),
) -> ApiResponse[list[MetadataRead]]:
    metadata = MetadataService(db).list(offset=offset, limit=limit)
    return ApiResponse(data=metadata)


@router.get("/{id}", response_model=ApiResponse[MetadataRead])
def get_metadata(
    id: UUID,
    db: Session = Depends(get_db),
) -> ApiResponse[MetadataRead]:
    metadata = MetadataService(db).get(id)
    return ApiResponse(data=metadata)


@router.put("/{id}", response_model=ApiResponse[MetadataRead])
def update_metadata(
    id: UUID,
    payload: MetadataUpdate,
    db: Session = Depends(get_db),
) -> ApiResponse[MetadataRead]:
    metadata = MetadataService(db).update(id, payload)
    return ApiResponse(message="Metadata updated.", data=metadata)


@router.delete("/{id}", response_model=ApiResponse[dict])
def delete_metadata(
    id: UUID,
    db: Session = Depends(get_db),
) -> ApiResponse[dict]:
    MetadataService(db).delete(id)
    return ApiResponse(message="Metadata deleted.", data={"id": str(id)})

File: frontend/src/services/metadata.js
/**
 * Metadata Service
 * 
 * API calls for:
 * - List metadata
 * - Get metadata by ID
 * - Create metadata
 * - Update metadata
 * - Delete metadata
 */

import apiClient from './api'

class MetadataService {
  /**
   * List all metadata with pagination
   */
  async listMetadata(offset = 0, limit = 100) {
    const response = await apiClient.get('/metadata', {
      params: { offset, limit },
    })
    return response.data.data
  }

  /**
   * Get metadata by ID
   */
  async getMetadata(id) {
    const response = await apiClient.get(`/metadata/${id}`)
    return response.data.data
  }

  /**
   * Create new metadata
   */
  async createMetadata(data) {
    const response = await apiClient.post('/metadata', data)
    return response.data.data
  }

  /**
   * Update metadata
   */
  async updateMetadata(id, data) {
    const response = await apiClient.put(`/metadata/${id}`, data)
    return response.data.data
  }

  /**
   * Delete metadata
   */
  async deleteMetadata(id) {
    const response = await apiClient.delete(`/metadata/${id}`)
    return response.data.data
  }

  /**
   * Search metadata
   */
  async searchMetadata(query) {
    const response = await apiClient.get('/metadata/search', {
      params: { q: query },
    })
    return response.data
  }
}

export default new MetadataService()
